nexusparser/tools/yaml2nxdl/yaml2nxdl.py

```diff
@@ -1,88 +1,114 @@
+import click
 
-    def parse(self):
-        #step2a: create an instantiated NXDL schema XML tree, begin with the header add XML schema/namespaces
-        attr_qname = etree.QName("http://www.w3.org/2001/XMLSchema-instance", "schemaLocation")
-        rt = etree.Element('definition', 
-                   {attr_qname: 'http://definition.nexusformat.org/nxdl/nxdl.xsd' },
-                   nsmap = {None: 'http://definition.nexusformat.org/nxdl/3.1', 
-                            'xsi': 'http://www.w3.org/2001/XMLSchema-instance'})
-        #step2b: user-defined attributes for the root group
-        if 'name' in self.yml.keys():
-            rt.set('name', self.yml['name'])
-            del self.yml['name']
-        else:
-            raise ValueError('ERROR: name: keyword not specified !')
-        pi = etree.ProcessingInstruction("xml-stylesheet", text='type="text/xsl" href="nxdlformat.xsl"')
-        rt.addprevious(pi)
-        #evaluate whether we handle an application definition, a contributed or base class
-        if 'category' in self.yml.keys():
-            if self.yml['category'] == 'application':
-                rt.set('category', 'application')
-                rt.set('extends', 'NXentry')
-            elif self.yml['category'] == 'contributed':
-                rt.set('category', 'contributed')
-                rt.set('extends', 'NXobject')
-            elif self.yml['category'] == 'base':
-                rt.set('category', 'base')
-                rt.set('extends', 'NXobject')
-            else:
-                raise ValueError('Top-level keyword category exists in the yml but one of these: application, contributed, base !')
-            del self.yml['category']
-            rt.set('type', 'group')
-        else:
-            raise ValueError('Top-level keyword category does not exist in the yml !')
-        #step2c: docstring
-        if 'doc' in self.yml.keys():
-            rt.set('doc', self.yml['doc'])
-            del self.yml['doc']
+
+def pretty_print_xml_given_root(root, output_xml):
+    """
+    Print formatted xml file with built-in libraries
+    """
+    xml_string = xml.dom.minidom.parseString(ET.tostring(root)).toprettyxml()
+    xml_string = os.linesep.join([s for s in xml_string.splitlines() if s.strip()])
+    with open(output_xml, "w") as file_out:
+        file_out.write(xml_string)
+
+
+@click.command()
+@click.option(
+    '--input-file',
+    help='The path to the input data file to read. (Repeat for more than one file.)'
+)
+def yaml2nxdl(input_file: str):
+
+    # step1
+    yml_appdef = read_application_definition(input_file)
+
+    # step2a
+    attr_qname = ET.QName("http://www.w3.org/2001/XMLSchema-instance", "schemaLocation")
+    rt = ET.Element(
+        'definition', {attr_qname: 'http://definition.nexusformat.org/nxdl/nxdl.xsd'},
+        nsmap={None: 'http://definition.nexusformat.org/nxdl/3.1',
+               'xsi': 'http://www.w3.org/2001/XMLSchema-instance'}
+    )
```

Comment:
The XML header part "nsmap" in this format seems to not be nexus-conform, at least the punx-tool "tree" function does not work with this header. The nexus nxdl files on github seem to have as header:
xmlns="http://definition.nexusformat.org/nxdl/3.1"
using this, it is recognized by punx

Reply:
Thank you for pointing out this discrepancy to nexus format. We indeed are moving from lxml to xml library and some functions are lacking, like the one generating the correct namespace. This issue has been now resolved and, once this pull request is closed, you will find the correct header in the xml generated file